0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,463 @@
+
+For the daprd sidecar to Scheduler Service communication, 
+![Scheduler APIs (SideCar Facing)](./resources/0004-BIRS-distributed-scheduler/sidecarToSchedulerComm.png)
+
+We will use the same exact protos from the Public Dapr API, but inside a **new**: `dapr/proto/scheduler/scheduler.proto`. However, the one difference will be in the `Job message`, where we will add:
+
+```proto
+    // Namespace for jobs to allow for multi tenancy.
+    string namespace = 5;
+
+    // The metadata associated with the job.
+    // The sidecar will create the unique `key` for storing data in the state store and pass the generated `key` along to the scheduler service for data lookup upon ‘trigger’ time later on. 
+    // The sidecar will also add metadata in order to know whether this job is registered for an actor. This is needed, as the routing mechanism for actors is different for the callback.
+    map<string,string> metadata = 6;
+```
+
+
+To allow for the daprd sidecar to register itself with the Scheduler Service upon startup, there will be a new `register` API in `dapr/proto/scheduler/scheduler.proto` with the following:
+
+```proto
+// This will allow for communication between the daprd sidecar and the Scheduler Service. (daprd sidecar -> Scheduler Service)
+service Scheduler {
+
+    ...<same user-facing protos defined above>...
+
+    // Allow for the daprd sidecar to register with a control plane Scheduler Service
+    rpc Register(RegisterRequest) returns (google.protobuf.Empty) {}
+}
+
+// RegisterRequest is the message to register a daprd sidecar with a control plane service. 
+message RegisterRequest {
+    // The id of the application (app_id). This is used as the key in etcd for the Scheduler Service.
+    string app_id = 1;
+
+    // The hostname of the daprd sidecar.
+    string hostname = 2;
+
+    // The port of the daprd sidecar.
+    int32 port = 3;
+}
+```
+
+Alternatively, there can be a streaming connection between the daprd sidecar and the Scheduler Service.
+
+###### Scheduler Service APIs
+
+In a **new** `dapr/proto/scheduler/scheduler_callback.proto`, the Scheduler Service will have a 'callback' to the app in which to fulfill the job trigger execution orchestration (ie. send back the data to the app at the time of its trigger). 
+
+```proto
+// This is intended for the Scheduler Service to call back to the daprd sidecar at the time of the job schedule... ie it's trigger time.
+service SchedulerCallback {
+
+    // Callback RPC for job schedule being at 'trigger' time
+    rpc TriggerJob(TriggerRequest) returns (google.protobuf.Empty) {}
+}
+
+// TriggerRequest is the message request to convey that the job is at its 'trigger' time. 
+message TriggerRequest {
+    // The job. 
+    Job job = 1; 
+
+    // The metadata associated with the job.
+    // This can contain the generated `key` for the optional state store when the daprd sidecar needs to lookup the entire data from a state store.
```

Comment:
This proposal was [shut down a few months ago](https://github.com/dapr/dapr/issues/6424) with the reasoning that it impacts performance negatively too much. Has the thinking changed about this?

Reply:
Data can be saved without limits in the default embedded db (etcd) and for flexibility users can use the state store of their choosing.